### PR TITLE
Change filesystem access to xdg-music only

### PR DESCRIPTION
**Pull request by Eonfge**
Not functional at the moment, requires some upstream attention

### CODE REVIEW

**Comment by Eonfge** [comment]
@bellegarde-c In line with Flatpak specs, I made a special Lollypop branch that has the sandbox more restrictive. I hope you can take a look at it, because at the moment there are two issues:

- The player crashes when you try to access an existing catalog directly
- The folder-selector is very rough, and doesn't seem to integrate with the GtkFileChooserNative

I have no intention of releasing this in its current state, but I do want you and others to be able to test it.

**Comment by bellegarde-c** [comment]
flatpak install --user https://dl.flathub.org/build-repo/29016/org.gnome.Lollypop.flatpakref

Done this but looks like Lollypop is not sandboxed here.

**Comment by bellegarde-c** [comment]
sandbox

**Comment by Eonfge** [comment]
Weird. I previously had issues opening the portal browser. Well, I'll run it through a few more tests in that case. 

The primary problem right now... is that this change is not backwards compatible. Users will have to re-select their music folders (other then dxg-music) to reimport their libraries. Is this worth the hassle?

Do you foresee any issues when we tighten up the sandbox permissions? 

@bilelmoussaoui do you have any comments concerning the sandboxing of Lollypop?

**Comment by Eonfge** [comment]
I actually ran into that myself. My version of Lollypop had a direct link to a non-xdg location. So over time, people will have to re-configure lollypop for the paths to properly migrate? I'll test that a few times, and we might have to hold off on it for a year or so, making sure that the fallout is not to big.

**Comment by Eonfge** [comment]
@bilelmoussaoui I've done some testing. As of right now, it will cause issues when we deploy this update, even for users who only downloaded Lollypop yesterday. 

What I've tested:
- Clean environment (Fedora 33, GNOME Boxes)
- Install lollypop flathub master
- Add custom folder with music (~/CustomMusicFolder/)
- Import, refresh, play for testing purposes
- Status: All good so far
- Install lollypop flathub NoMoreHost
- import, refresh
- Status: Files can't be found

I had to re-select my music library and I had to reimport it. The current version of Lollypop and GtkFileChooserNetive, don't use a a /var/run/ mount when selecting files, if they don't have to. As such, when updating lollypop it will be unable to access the files in ~/CustomMusicFolder/

So, what's the plan? I do want to improve the security in the future

**Comment by bellegarde-c** [comment]
@Eonfge Can I do something from Lollypop POV ?

**Comment by bilelmoussaoui** [comment]
Nothing can be done regarding the migration as the files/directories you grant an application the permission to open/read/write are stored in the flatpak documents store and are added once you open a file using the portal (the native file dialog). The directories has to be re-added so that Lollypop has actually the permission granted to open/read those files instead of "having permissions to everything"

**Comment by bilelmoussaoui** [comment]
> @Eonfge Can I do something from Lollypop POV ?

what would be nice is to migrate all file choosers to native ones as a first step

**Comment by Eonfge** [comment]
@bellegarde-c I was thinking about that. Would it be possible to add a `-DFlatpak=true` flag to meson? 

You can then show a message that informs users about their folders and permissions when they start Lollypop and a folder is missing. If a folder is missing, you can tell users like like "Sandboxing good. Sandboxing WIP. Please re-import". 

The comment of @bilelmoussaoui is certainly true, making sure all file choosers are up-to-date is important, but I think we can't escape the realization that enforcing sandboxing does require a minor action from the user.

**Comment by Eonfge** [comment]
@bellegarde-c Would it be possible to provide some user-notification when they start the Flatpak version first?

If you could include such a message, I can combine it with this change. Then, when users update Lollypop to a more sandboxed version, they'll be informed about the change. The flag and notification could be temporary, because as time progresses all people will migrate. After a few months, we could remove it.

**Comment by Eonfge** [comment]
Screenshot from 2020-12-20 13-26-48

@bellegarde-c Would such a one-time notice be possible? There is not really a way to migrate users, but I do want to remove the host-access from this Flatpak. 

It should only show on startup when the project is build with '-DFlatpak=true' and it should be a one-time thing. After a few months, we can remove it again since then all users would have migrated. In the backend, we can use a simple g-schema bool to check if the user has ever clicked on continue, which would prevent it from showing ever again.

**Comment by bellegarde-c** [comment]
Here we go: https://gitlab.gnome.org/World/lollypop/-/commit/d7bc69b67ffb045bde51dd3c57163113fcd23001

Don't want technical information. Sandboxing is not a user word :-)

migration1

migration2


It's a one time migration, so we need to test this before releasing on flathub ;) 

To test again, in dconf-editor:

dconf

**Comment by Eonfge** [comment]
> Don't want technical information. Sandboxing is not a user word :-)

And that's why you make a kickass music player, and I bother packaging it :innocent: 

I'll be doing some testing using this version and if all goes according to plan, then I'll change the sandboxing settings with the next release.

**Comment by Eonfge** [comment]
Well, first tests positive, works both with my existing player and in a VM. 

Do have a few nit-picks though:
- The dialogue is not necessary for users who install Lollypop fresh
  - On a fresh start following the wizard, the settings don't show the xdg-music folder pre-selected, but it does snow it pre-selected once you open the settings menu afterwards
- Clicking cancel in the first dialogue still opens preferences

Unrelated:
- The pop-up about 'Automatically download metadata?' doesn't actually enable network access, thus being non-functional

**Comment by bellegarde-c** [comment]
>The dialogue is not necessary for users who install Lollypop fresh

Does it mean that XDG music dir is available by default ?

**Comment by Eonfge** [comment]
> > The dialogue is not necessary for users who install Lollypop fresh
> 
> Does it mean that XDG music dir is available by default ?

That seems completely reasonable for a music player, so I've enabled it by default. Lollypop can also access GVFS in case that people host their music on a network.

**Comment by bellegarde-c** [comment]
Do you think there is a simple way to test wizard is needed ?

**Comment by Eonfge** [comment]
> Do you think there is a simple way to test wizard is needed ?

Not sure... @bilelmoussaoui Do you know if it's possible to detect a dxg-music folder from the inside?

There is of cause a simple workaround: Only show the message if there is already music imported.

**Comment by bellegarde-c** [comment]
@bilelmoussaoui What I want is detecting configured folders in Lollypop with access denied by sandbox.

**Comment by bellegarde-c** [comment]
> * The pop-up about 'Automatically download metadata?' doesn't actually enable network access, thus being non-functional

This one is working here. Do you mean that lollypop is not downloading artwork or that the setting is not updated ?

**Comment by bilelmoussaoui** [comment]
> @bilelmoussaoui What I want is detecting configured folders in Lollypop with access denied by sandbox.

You can query the documents portal and see which permissions your app has for a specific folder/file

**Comment by Eonfge** [comment]
> > * The pop-up about 'Automatically download metadata?' doesn't actually enable network access, thus being non-functional
> 
> This one is working here. Do you mean that lollypop is not downloading artwork or that the setting is not updated ?

It would be the setting then. This is what I see after I click agree:
Screenshot from 2020-12-27 14-28-35

**Comment by bellegarde-c** [comment]
It modifies the background data, it's another issue :D

**Comment by bellegarde-c** [comment]
Updated Lollypop to handle migration only if needed.

**Comment by Eonfge** [comment]
I've successfully tested the following two scenarios:
- Do a fresh installation, select custom folder
- Do a fresh installation, use xdg-music
- Have an existing library, have music in a custom folder, re-import folder required
- Have an existing library, have music in xdg-music and a custom folder, xdg-folder is instantly available, custom folder must be re-imported
- Have an existing library, have music in xdg-music only, no re-import required

### Bugs

- **UI Bug**: in scenario 2, the dxg-folder is not visible when I open the preferences menu for the first time. It only becomes visible if I then clos it and re-open it.
- **Underwater Bug**: In all scenarios, I get a message that Last FM and Listen Brains can't find the appropriate files. 
[log.txt](https://github.com/flathub/org.gnome.Lollypop/files/5745335/log.txt)

### In summary
I think we have this fully working :smile: I would invite others to try it, but I haven't found any issues.

**Comment by bellegarde-c** [comment]
UI bug was a Lollypop bug, fixed ;)

**Comment by Eonfge** [comment]
Well, all seems well. Let me know when there is a new Lollypop release and I'll release it with the improved sandboxing.

**Comment by bellegarde-c** [comment]
Time to merge this, 1.4.8 is out.

**Comment by Eonfge** [comment]
Closing this branch in favour of #122 so we don't have a whole series of surplus comments on Git
